apps/web/src/components/Messages/Composer.tsx

```diff
@@ -1,45 +1,129 @@
 import useWindowSize from '@components/utils/hooks/useWindowSize';
-import { ArrowRightIcon } from '@heroicons/react/outline';
+import { ArrowRightIcon, PhotographIcon } from '@heroicons/react/outline';
 import { Mixpanel } from '@lib/mixpanel';
+import { uploadFileToIPFS } from '@lib/uploadToIPFS';
 import { t, Trans } from '@lingui/macro';
+import type { ContentTypeId } from '@xmtp/xmtp-js';
+import { ContentTypeText } from '@xmtp/xmtp-js';
 import { MIN_WIDTH_DESKTOP } from 'data/constants';
-import type { FC } from 'react';
-import { useEffect, useState } from 'react';
+import type { ChangeEvent, FC } from 'react';
+import { useEffect, useRef, useState } from 'react';
 import toast from 'react-hot-toast';
+import {
+  useAttachmentCacheStore,
+  useAttachmentStore
+} from 'src/store/attachment';
 import { useMessagePersistStore } from 'src/store/message';
 import { MESSAGES } from 'src/tracking';
 import { Button, Input, Spinner } from 'ui';
+import type {
+  Attachment,
+  RemoteAttachment
+} from 'xmtp-content-type-remote-attachment';
+import {
+  AttachmentCodec,
+  ContentTypeRemoteAttachment,
+  RemoteAttachmentCodec
+} from 'xmtp-content-type-remote-attachment';
+
+import AttachmentView from './AttachmentView';
 
 interface ComposerProps {
-  sendMessage: (message: string) => Promise<boolean>;
+  sendMessage: (content: any, contentType: ContentTypeId) => Promise<boolean>;
   conversationKey: string;
   disabledInput: boolean;
 }
 
+const AttachmentComposerPreview = ({
+  onDismiss,
+  attachment
+}: {
+  onDismiss: () => void;
+  attachment: Attachment;
+}): JSX.Element => {
+  return (
+    <div className="relative ml-12 inline-block rounded pt-6">
+      <Button className="absolute right-4 top-4" size="sm" onClick={onDismiss}>
+        Remove
```

Comment:
Added this, but generally i18n's `<Trans>` is only needed for complex elements e.g. if you're doing some sort of string interpolation; not sure why it'd be needed in this case instead of the `t` function (maybe this package uses it differently). Added it anyway to align with the pattern used throughout the app.